File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        count = [k, 0]
        self.dfs(root, count)
        return count[1]

    def dfs(self, root, count) -> None:
        if not root or count[0] == 0:
            return
        
        self.dfs(root.left, count)
        
        count[0] -= 1
        if count[0] == 0:
            count[1] = root.val
            return
        self.dfs(root.right, count)
        

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null) return new ArrayList<>();
        
        List<Integer> res = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()) {
            int count = queue.size();
            TreeNode node = null;
            for(int i = 0; i < count; i++) {
                node = queue.poll();
                if(node.left != null) queue.offer(node.left);
                if(node.right != null) queue.offer(node.right);   
            }
            res.add(node.val);
        }
        return res;
    }
}
